prog/dftb+/lib_dftb/encharges.F90

```diff
@@ -8,28 +8,442 @@
+  type, extends(TEeqParam) :: eeqInput
+
+    !> Net charge of the system.
+    real(dp) :: nrChrg
+
+    !> Parameter for Ewald summation.
+    real(dp) :: parEwald
+
+    !> Ewald tolerance
+    real(dp) :: tolEwald
+
+    !> Cutoff for real-space summation under PBCs
+    real(dp) :: cutoff
+
+  end type eeqInput
+
+
+  !> Container containing all calculation data for the EEQ model
+  type :: TEeqCont
+
+    !> number of atoms
+    integer :: nAtom
+
+    !> EEQ parametrisation
+    type(TEeqParam) :: param
+
+    !> Net charge of the system.
+    real(dp) :: nrChrg
+
+    !> Cutoff for real-space summation under PBCs
+    real(dp) :: cutoff
+
+    !> lattice vectors if periodic
+    real(dp) :: latVecs(3, 3)
+
+    !> Volume of the unit cell
+    real(dp) :: vol
+
+    !> evaluate Ewald parameter
+    logical :: tAutoEwald
+
+    !> Ewald tolerance
+    real(dp) :: tolEwald
+
+    !> Parameter for Ewald summation.
+    real(dp) :: parEwald
+
+    !> is this periodic
+    logical :: tPeriodic
+
+    !> Contains the points included in the reciprocal sum.
+    !> The set should not include the origin or inversion related points.
+    real(dp), allocatable :: recPoint(:, :)
+
+    !> are the coordinates current?
+    logical :: tCoordsUpdated
+
+    !> electrostatic energy
+    real(dp), allocatable :: energies(:)
+
+    !> force contributions
+    real(dp), allocatable :: gradients(:, :)
+
+    !> stress tensor
+    real(dp), allocatable :: stress(:, :)
+
+    !> partial charges
+    real(dp), allocatable :: charges(:)
+
+    !> derivative of partial charges w.r.t. coordinates
+    real(dp), allocatable :: dqdr(:, :, :)
+
+    !> derivative of partial charges w.r.t. strain deformations
+    real(dp), allocatable :: dqdL(:, :, :)
+
+  contains
+
+    !> initialize container from input
+    procedure :: initialize
+
+    !> update internal store of coordinates
+    procedure :: updateCoords
+
+    !> update internal store of lattice vectors
+    procedure :: updateLatVecs
+
+    !> return energy contribution
+    procedure :: getEnergies
+
+    !> return energy contribution
+    procedure :: getCharges
+
+    !> return force contribution
+    generic :: addGradients => addGradientsEnergy, addGradientsCharges
+
+    !> return force contribution for electrostatic energy
+    procedure, private :: addGradientsEnergy
+
+    !> return force contribution for charge derivatives
+    procedure, private :: addGradientsCharges
+
+    !> return stress tensor contribution
+    generic :: addStress => addStressEnergy, addStressCharges
+
+    !> return stress tensor contribution for electrostatic energy
+    procedure, private :: addStressEnergy
+
+    !> return stress tensor contribution for charge derivatives
+    procedure, private :: addStressCharges
+
+    !> cutoff distance in real space for EEQ
+    procedure :: getRCutoff
+
+  end type TEeqCont
+
+
 contains
 
+
+  subroutine initialize(this, input, withEnergies, withCharges, nAtom, latVecs)
```

Comment:
Are there any arguments against using a `function` and overloading the constructor:

```fortran
interface TEqCont
  module procedure :: initialize
end interface TEqCont
```

Reply:
The alternative is to keep the constructor type bound and call it from the pure type constructor, like you would call a type bound deconstructor from the finalizer. If you inherit from the class, you have to overwrite both constructor and deconstructor and call back to its parent, which would make more sense to me.

To be fair, I actually don't see a need to have the classes here extensible, it is just more convenient to have all the procedures type bound to avoid polluting the global namespace.